### ISSUE TITLE
Getting count of records in child table

### ISSUE THREAD

**Issue by evanverneyfink** [dotnet/efcore#5413]
Hopefully I'm putting this in the right place. It certainly seems like an issue to me, although it's also possible I'm going about this the wrong way.

I'm working with the EntityFramework.Core and EntityFramework.MicrosoftSqlServer 7.0.0-rc1-final packages, and I'm trying to do something through code that looks like the following SQL:

```
SELECT COUNT(*)
FROM [UploadPackage] [up]
INNER JOIN [RawClip] [rc] on [up].[Id] = [rc].[PackageId]
WHERE [up].[UserId] IN (1,2,3)
```

I've tried the following:

Option 1:

```
DbContext.UploadPackages.Include(up => up.Clips).Where(up => userIds.Contains(up.UserId)).SelectMany(u => u.Clips).CountAsync()
```

Option 2:

```
DbContext.RawClips.Include(rc => rc.Package).Where(rc => userIds.Contains(rc.Package.UserId)).CountAsync()
```

The relationships are configured using the Fluent API like so:

```
modelBuilder.Entity<UploadPackage>().HasMany(up => up.Clips).WithOne(rc => rc.Package);
modelBuilder.Entity<RawClip>().HasOne(rc => rc.Package).WithMany(up => up.Clips);
```

Option 1 generates the following SQL:

```
SELECT [up].[Id], [up].[AssetId], [up].[FileName], [up].[FileSize], [up].[PackageId], [up].[Path], [up].[ProxyUrl], [up].[StatusDetails], [up].[StatusId], [up].[ThumbnailUrl], [up].[UploadCompletedDateTime], [up].[UploadStartedDateTime], [r].[Id], [r].[AssetId], [r].[FileName], [r].[FileSize], [r].[PackageId], [r].[Path], [r].[ProxyUrl], [r].[StatusDetails], [r].[StatusId], [r].[ThumbnailUrl], [r].[UploadCompletedDateTime], [r].[UploadStartedDateTime], [up].[Id]
FROM [RawClip] AS [up]
CROSS JOIN [RawClip] AS [r]
WHERE [up].[UserId] IN (2, 3, 4, 1)
```

As you can see, it's joining the RawClip table to itself. On one side of the join, it's giving the table the alias "[up]" which seems to be a reference to UploadPackage, even though the table it's aliasing is actually the RawClip table. This ultimately throws an error because of the where clause trying to filter on the column "[up].[UserId]", which exists on the UploadPackage table, but not the RawClip table.

Option 2 fails without generating any SQL. Here's the error with stack trace:

```
InvalidOperationException: Sequence contains more than one element
System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
Microsoft.Data.Entity.Query.EntityQueryModelVisitor.<>c__DisplayClass79_0`1.<BindMemberExpression>b__0(IEnumerable`1 ps, IQuerySource qs)
Microsoft.Data.Entity.Query.EntityQueryModelVisitor.BindMemberExpressionCore[TResult](MemberExpression memberExpression, IQuerySource querySource, Func`3 memberBinder)
Microsoft.Data.Entity.Query.EntityQueryModelVisitor.BindMemberExpression[TResult](MemberExpression memberExpression, IQuerySource querySource, Func`3 memberBinder)
Microsoft.Data.Entity.Query.EntityQueryModelVisitor.BindMemberExpression(MemberExpression memberExpression, Action`2 memberBinder)
Microsoft.Data.Entity.Query.ExpressionVisitors.Internal.RequiresMaterializationExpressionVisitor.VisitMember(MemberExpression memberExpression)
System.Linq.Expressions.MemberExpression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
Microsoft.Data.Entity.Query.ExpressionVisitors.ExpressionVisitorBase.Visit(Expression expression)
Remotion.Linq.Clauses.ResultOperators.ContainsResultOperator.TransformExpressions(Func`2 transformation)
Remotion.Linq.QueryModel.TransformExpressions(Func`2 transformation)
Microsoft.Data.Entity.Query.ExpressionVisitors.Internal.RequiresMaterializationExpressionVisitor.VisitSubQuery(SubQueryExpression subQueryExpression)
Remotion.Linq.Clauses.Expressions.SubQueryExpression.Accept(ExpressionVisitor visitor)
System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
2000 
Microsoft.Data.Entity.Query.ExpressionVisitors.ExpressionVisitorBase.Visit(Expression expression)
Remotion.Linq.Clauses.WhereClause.TransformExpressions(Func`2 transformation)
Remotion.Linq.QueryModel.TransformExpressions(Func`2 transformation)
Microsoft.Data.Entity.Query.ExpressionVisitors.Internal.RequiresMaterializationExpressionVisitor.FindQuerySourcesRequiringMaterialization(QueryModel queryModel)
Microsoft.Data.Entity.Query.QueryCompilationContext.FindQuerySourcesRequiringMaterialization(EntityQueryModelVisitor queryModelVisitor, QueryModel queryModel)
Microsoft.Data.Entity.Query.EntityQueryModelVisitor.CreateAsyncQueryExecutor[TResult](QueryModel queryModel)
Microsoft.Data.Entity.Storage.Database.CompileAsyncQuery[TResult](QueryModel queryModel)
Microsoft.Data.Entity.Query.Internal.QueryCompiler.<>c__DisplayClass19_0`1.<CompileAsyncQuery>b__0()
Microsoft.Data.Entity.Query.Internal.CompiledQueryCache.GetOrAddAsyncQuery[TResult](Object cacheKey, Func`1 compiler)
Microsoft.Data.Entity.Query.Internal.QueryCompiler.CompileAsyncQuery[TResult](Expression query)
Microsoft.Data.Entity.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
Microsoft.Data.Entity.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
Microsoft.Data.Entity.EntityFrameworkQueryableExtensions.ExecuteAsync[TSource,TResult](MethodInfo operatorMethodInfo, IQueryable`1 source, CancellationToken cancellationToken)
Microsoft.Data.Entity.EntityFrameworkQueryableExtensions.CountAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
```

Am I doing something wrong? Is there a better way to achieve what I'm trying to do? It seems like a fairly simple use case, so I imagine there has to be some way to make this work...

**Comment by evanverneyfink** [comment]
Is there something about my post that's causing it to be ignored?

This seems like a legitimate bug to me. I was able to put together a very simple project that recreates the problem:

https://github.com/evanverneyfink/EFChildCountIssue.

**Comment by rowanmiller** [comment]
@evanverneyfink we just don't triage incoming issues everyday. We get a very large number of incoming issues on this repo so we have a process to deal with it. We'll go thru the new issues on Monday.

**Comment by evanverneyfink** [comment]
Fair enough. Didn't mean to be pushy, I just saw other issues with responses and thought perhaps I'd done something incorrectly. Thanks for the response!

**Comment by smitpatel** [comment]
@evanverneyfink -
I took the repo you provided for repro. Since it was in RC1 I pulled out model etc in new project using RC2 packages.
Using the model, DbContext & seed data provided in `TestDbContext`
````
public static void Main(string[] args)
{
    using (var context = new TestDbContext())
    {
        context.Database.EnsureDeleted();
        context.Database.EnsureCreated();
        context.EnsureSeedData();
    }


    using (var context = new TestDbContext())
    {
        var userIds = new List<int> {1, 2, 3, 4};

        var count =
            context.Packages.Where(pkg => userIds.Contains(pkg.UserId)).SelectMany(pkg => pkg.Files).Count();
        Console.WriteLine(count);
    }
}
````
Output is 27
Generated query is
````
SELECT COUNT(*)
FROM [Packages] AS [pkg]
INNER JOIN [PackageFiles] AS [pkg.Files] ON [pkg].[Id] = [pkg.Files].[PackageId]
WHERE [pkg].[UserId] IN (1, 2, 3, 4)
````
Which is same as what you are looking for. Upgrading to RC2 packages will solve your issue.
Another thing, using `SelectMany` on navigation will create the join, you do not need to use `Include` on the same navigation.

**Comment by maumar** [comment]
duplicate of #3889, fixed in 7d27bf6527391cd46dbb09210f7e40f7998e549e, closing

**Comment by evanverneyfink** [comment]
Thanks guys. I'll upgrade to RC2.
